Prometheus: Configuration page overhaul

bohandley opened the pull request:
**What is this?**
[Config Overhaul Stage 1](https://docs.google.com/document/d/1WqEbJvInT_94xA4M-pATMYypnjGIs3dbwNbEvsqGuF8/edit#) This is a first stage of the [Prometheus Config Overhaul Epic](https://github.com/grafana/grafana/issues/65877) as driven by G10 Getting Started and the Config Overhaul Working Group. These changes were described and shared to me by @josiahg, @catherineymgui and @gtk-grafana 

- [x] See discussion below with @torkelo and @catherineymgui, we are adding a prom label and interactive tooltip to the URL input in DatasourceHttpSettting.ts instead of creating a new custom bare minimum section. The bare minimum section will be added in stage 2 once the auth component has been refactored and the styles are consistent. For now, we are incrementally moving toward the config overhaul working group goal of making the UI more clear with Prom label/adding doc tooltip for setting up a Prometheus datasource.
<img width="528" alt="Screenshot 2023-04-26 at 3 28 41 PM" src="https://user-images.githubusercontent.com/25674746/234682482-f5235f21-f305-4d95-adc8-39e54c65934c.png">

  - auth will be a new component added in stage 2 and design consistency will be updated
  - Connection section will be created in stage 2 with the auth and design with consistent padding will be completed.
  
- [x] organizing other config elements into categories to not clutter the page (maybe hide additional elements)
<img width="345" alt="Screenshot 2023-04-17 at 3 04 33 PM" src="https://user-images.githubusercontent.com/25674746/232586015-70da24b2-8e84-4012-8204-fd5bdbb5309f.png">

- [x] Save & Test: adding test backend datasource health check to improve error handling(url and auth)
  - Examples
  - [x] implement `/-/healthy` api for prometheus types that support this api
![Screenshot 2023-04-16 at 6 58 17 PM](https://user-images.githubusercontent.com/25674746/232586245-05997fb1-e0f1-4dcc-96e7-a00e00b4dc4b.png)

  - [x] Url error with timeout
  <img width="488" alt="Screenshot 2023-04-17 at 2 41 27 PM" src="https://user-images.githubusercontent.com/25674746/232588563-a9b96609-3a11-4a5d-95f9-6689131d53b9.png">
  <img width="387" alt="Screenshot 2023-04-17 at 2 41 47 PM" src="https://user-images.githubusercontent.com/25674746/232588566-1ae0ab1b-2094-4f5a-bdc7-d4c9caa5ea91.png">
  <img width="735" alt="Screenshot 2023-04-17 at 2 55 22 PM" src="https://user-images.githubusercontent.com/25674746/232588877-da793526-d7f5-4616-a6ae-e39305fefe8b.png">

  - [x] Simple url error 
  <img width="245" alt="Screenshot 2023-04-17 at 2 42 30 PM" src="https://user-images.githubusercontent.com/25674746/232588662-4c5bcaae-7e6e-43f1-be0c-c01141bf18c0.png">
  <img width="294" alt="Screenshot 2023-04-17 at 2 50 38 PM" src="https://user-images.githubusercontent.com/25674746/232588664-deb94840-bbf6-4509-a7cc-d2cf59ad7c2a.png">

  - [x] fallback to a query for instances that do not implement the `/-/healthy` api
![Screenshot 2023-04-16 at 6 59 21 PM](https://user-images.githubusercontent.com/25674746/232586541-d6ca266a-693d-4ab9-afab-a240aa0107f8.png)

  - [x] incorrect url for instance that does not implement the `/-/healthy` api
![Screenshot 2023-04-16 at 8 53 08 PM](https://user-images.githubusercontent.com/25674746/232587098-a2266bea-3d6a-4079-b1f8-9f12313d03f7.png)

  - [x] incorrect authentication for instance that does not implement the `/-/healthy` api
![Screenshot 2023-04-16 at 9 02 34 PM](https://user-images.githubusercontent.com/25674746/232587246-fd110cd1-9204-4696-85e0-4e7e6c490142.png)




- [x] adding links to  documentation
- [x] updating tooltips for config elements
- [x] removing legacy form inputs

This work aligns with the Docs Overhaul I am working on with Larissa Wandzura on the docs team. We are updating the Prometheus Configuration Docs for the overhaul. https://grafana.com/docs/grafana/latest/datasources/prometheus/#configure-the-data-source

The UI suggestions, feature updates and organization plans are listed in this doc which has assistance from @itsmylife and others mentioned above. https://docs.google.com/document/d/1Pw0l30ijhBAnXqm5IFPU8jPEtiL5ZtW29OQhE-dU5l8/edit#

**Why do we need this feature?**
G10 Getting Started: Prometheus research and the Config working group uncovered issues with datasource configuration pages in Grafana. People can get stuck setting up datasources. 90% of people click save and test but 4% are successful on start. The work suggested will help improve people's success in setting up datasources.

**Who is this feature for?**
Grafana users who are setting up a Prometheus datasource.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/65877
Fixes https://github.com/grafana/grafana/issues/65970
Fixes https://github.com/grafana/grafana/issues/66829

**Special notes for your reviewer:**
This is a first stage. The tooltips may need more updating, the links to docs need updated by the work done with Larissa, the auth component and configurable header component are being worked on by @grafana/enterprise-datasources, and error handling from testing the datasource needs updated (another PR for creating a backend health check which prometheus does not have).

The previous config page is here:
<img width="381" alt="Screenshot 2023-04-08 at 2 41 00 PM" src="https://user-images.githubusercontent.com/25674746/230737681-c7b9d4d5-b8a8-41ed-a54b-5ce9538b8a78.png">

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

bohandley committed:
organize layout, make design uniform, add doc link

bohandley committed:
fix e2e test

bohandley committed:
move overhauled config parts into prometheus code

catherineymgui reviewed (dismissed):
@josiahg and I took a look and had comments

UI:

- [x] 32px between the section headers
- [x] Section descriptions - use secondary grey colour
- [x] Subheaders under additional settings - use h6
- [x] Move the toggle in outside of the field in Manage alerts via Alerting UI and Disable metrics lookup
  - [X] *Brendan - most are done but need to figure out a nice way to do Alerting
- [x] It seems most of the form fields are using a legacy input field (which is why certain fields have different tooltip colours). Could you double check this
  - [X] *Brendan - most are done in PromSettings.ts and DatasourceHttpSettingsOverhaul.ts but there might still be some and I need to poke around some more. 

  Copy:
  
- [X] “See the documentation” link opens in the same tab instead of new tab, rips you out of Grafana. Also there’s an extra period
- [x] See documentation section — use secondary grey
- [ ] See documentation section — Note: We are adding an additional component (https://github.com/grafana/plugin-ui/pull/64) — this is in plugin-ui now, but will be added to core. I don’t know if you want to remove this for now
  - [ ] *Brendan - holding off on this one for now until it's in core grafana
- [x] Potential renaming of groups: Prometheus Behavioural —> Interval Behaviour, UI —> Query Editor, Advanced —> Other
- [x] While we are not perfectly consistent, prefer “data source” to “datasource” throughout the page
- [x] URL -> “Prometheus Server URL”
- [x] Tooltip: “This URL must be accessible from the Grafana server”
- [x] Add tooltip for “manage alerts via Alerting UI” — Matias made one here https://github.com/grafana/grafana/pull/66216#event-8965241606
- [x] Prometheus Behavioural - tooltips don’t help me determine this (where do i look in Prom to find this info?) they also don’t let me know if there are any consequences if I do it wrong - if I have my Prom set for 10s and i set this config to 60s what happens?
- [x] “Default editor” tooltip needs updated to - “Set default editor option for all users of this data source. If no option was selected, the default editor will be the "builder".”
- [x] Can we just set the default to Builder in the config UI and delete the second sentence as well?
- [x] Prometheus type - no information about implications if we do this wrong
- [x] Custom query parameters - what is the reason you would want to do this? (Contrast with HTTP method tooltip which is quite good and explains implications)
- [x] Exemplars - Internal Link - no tooltip, what does this setting do?
- [x] Would be good to ask Larissa to take a look to ensure consistency in content, and tooltips

bohandley committed:
update tooltips with doc links

bohandley committed:
clean component styles for section padding, top and bottom 32px

bohandley committed:
make additional settings subsection headers h6

bohandley committed:
use secondary gray for section descriptions

bohandley committed:
Merge branch 'main' into bohandley/prometheus-config-overhaul

bohandley committed:
fix merge issues

bohandley committed:
change inlineswitch to switch only in prom settings because the other components are shared

bohandley committed:
remove legacy formfield and input, replace with inlinefield and input from grafana-ui

bohandley committed:
find more formfield and UI design fixes like changing inlineformlabel to inlinefield

bohandley commented:
Need to work on the following inputs that may be shared across other datasources. I need to either recreate them with the right design(InlineField over FormField and Switch over InlineSwitch)
<img width="854" alt="Screenshot 2023-04-12 at 8 56 08 PM" src="https://user-images.githubusercontent.com/25674746/231619086-2e09312b-8fd3-4d12-8ad3-91d8de224d7b.png">
<img width="297" alt="Screenshot 2023-04-12 at 8 56 16 PM" src="https://user-images.githubusercontent.com/25674746/231619089-11df7efb-676e-429b-9ddb-e5eba925e3cf.png">

Need to recreate validation events for `Scrape Interval`, `Query Timeout`, and `Query overlap window` for incremental querying because only legacy Input component accepts validation events. There should be another way to validate the Input component value from grafana-data.

bohandley committed:
remove unused inline form label

bohandley committed:
replace legacy duration validations with <FieldValidationMessage>

bohandley committed:
fix styles secondary gray from theme

bohandley committed:
change language, headings and datasource -> data source

bohandley committed:
update alert setting styles with new component

bohandley committed:
update prom url heading and tooltip

bohandley committed:
update default editor tooltip and set builder as default editor

bohandley committed:
update interval tooltip

bohandley committed:
update prom type tooltip

bohandley committed:
update custom query params tooltip

bohandley committed:
update exemplar internal link tooltip

bohandley committed:
fix inline form styling inconsistency

sasklacz commented:
@bohandley as for the error handling, for now please hold this off as we plan to rewrite one of the plugins later this month to show how this could be done taking into account all of our previous ideas and conclusions. So realistically we can start shipping those plugins sometime in May.

itsmylife commented:
@bohandley Following image is not related to your work but we don't really show the details. Maybe in this PR, you wanna address this?
![image](https://user-images.githubusercontent.com/820946/232037964-7d9a0a17-826f-4fb2-a09c-8bd70072c81f.png)

itsmylife commented:
I am using a wide monitor and the config page looks like this on my screen. In small monitors, they go line by line but this way I cannot determine which toggle is for which setting. It is kinda a bad experience for me. 
![image](https://user-images.githubusercontent.com/820946/232038136-b13888b7-179c-467f-b4ff-c5254d0f6548.png)

itsmylife commented:
Also labels and input fields are not aligning. I thought this can be also addressed in this PR. If that's not the intention we can open another issue to address this. 
![image](https://user-images.githubusercontent.com/820946/232039618-cf5c19ad-2261-42d5-9911-c9404d1a0faf.png)

itsmylife commented:
Also another aligning issue picture 
![image](https://user-images.githubusercontent.com/820946/232040810-300a5ac8-69ee-413b-adef-c8df5abd7ba4.png)

itsmylife commented:
So maybe @catherineymgui can be the right person for this kind of question. 
Referring the thread https://raintank-corp.slack.com/archives/C0F5GDR4P/p1681224647575209

cletter7 commented:
Yes, I am currently working on the Auth component designed by @catherineymgui that should replace the `DataSourceHttpSettings`

bohandley commented:
Thank you @catherineymgui @sasklacz @cletter7!

@yesoreyeram and @itsmylife Catherine, Kuba, Andriy and myself are meeting today to sync up about this work. Much of this PR will be pulled back/removed, so @josiahg there will be major design differences. I will update the things I can based on the meeting. Thank you everyone!

bohandley committed:
allow for using the DataSourceHTTPSettings component without the connection string

bohandley committed:
remove overhaul component, re-use dshtttps comp, and use connection input in config editor

bohandley committed:
make tooltips interactive to click links

bohandley committed:
consistent label width across the elements we can control for now, fix exemplar switch

bohandley commented:
I removed the duplicate efforts @yesoreyeram @cletter7 , I made the elements that we have control over have consistent label width @itsmylife, I added interactive tooltips to click on links in tooltips @josiahg. This will be stage one according to our meeting earlier today. Stage 2 will be where I use the auth component that is being developed in plugins-ui and wrap that change behind a feature toggle. https://docs.google.com/document/d/1WqEbJvInT_94xA4M-pATMYypnjGIs3dbwNbEvsqGuF8/edit?usp=sharing

@josiahg after our meeting I was able to get the save and test to return a false positive so I will be working on the health check for the backend. This will be in another PR though because I will need @grafana/observability-metrics to weigh in on the depth of what the health check will test.

Thanks all, have a great weekend!

bohandley committed:
make connection url a component

bohandley committed:
refactor onBlur validation

bohandley committed:
remove unused component

bohandley committed:
add tests for config validations

bohandley committed:
add more meaningful health check

bohandley committed:
fix e2e test

bohandley committed:
fix e2e test

bohandley committed:
fix e2e test

bohandley committed:
add error handling for more url errors

bohandley commented:
Adding health check to this PR so that we can get it in before G10 cutoff on May 1

bohandley committed:
remove unnecessary conversion

bohandley committed:
health check tests

itsmylife committed:
Clean up the health check

bohandley commented:
The simple URL check is not working currently, need to fix this. There are two responses for this, either a timeout or a false positive where a response is sent but the response is a redirect or some html from the network.

itsmylife committed:
health check unit tests

itsmylife committed:
health check unit tests improved

bohandley committed:
Merge branch 'main' into bohandley/prometheus-config-overhaul

bohandley committed:
make pretty for drone

bohandley committed:
lint check go

bohandley committed:
lint check go

bohandley committed:
add required attr to connection component

catherineymgui commented:
@bohandley 

Let me know if these error / success messages are accurate:

Success

- Successfully saved the configuration and queried the Prometheus API.

Timeout
- Error querying resource: Get “[http://this./-/healthy”:](http://this./-/healthy%E2%80%9D:) dial tcp 143.244.220.150.80: io timeout  - Your configuration has been saved but there was a timeout. Try saving again.

Simple URL
- Your configuration has been saved but there is an error. Double check the URL formatting.

Cannot connect to the URL
- Your configuration has been saved but there is an error. Double check the URL formatting.

  Incorrect URL / 404
- Your configuration has been saved but there is an error. Double check the URL formatting .

Incorrect authentication / 401
- 401 Unauthorized - Your configuration has been saved but there is an error. Double check the authentication.

gtk-grafana commented on public/app/plugins/datasource/prometheus/configuration/ConfigEditor.test.tsx:
```diff
@@ -0,0 +1,48 @@
+    ${'1h'}  | ${true}
+    ${'1m'}  | ${true}
+    ${'1s'}  | ${true}
+    ${'1y'}  | ${true}
```
will multiple duration types work? i.e `1h 5m 2s`?

bohandley replied:
Good catch! No it does not, this was a custom prom settings duration validation
```
export const promSettingsValidationEvents = {
  [EventsWithValidation.onBlur]: [
    regexValidation(
      /^$|^\d+(ms|[Mwdhmsy])$/,
      'Value is not valid, you can use number with time unit specifier: y, M, w, d, h, m, s'
    ),
  ],
};
```

This is different from the `isValidDuration` duration check that you imported for incremental querying
```
/**
 * isValidDuration returns true if the given string can be parsed into a valid Duration object, false otherwise
 *
 * @param durationString - string representation of a duration
 *
 * @public
 */
export function isValidDuration(durationString: string): boolean {
  for (const value of durationString.trim().split(' ')) {
    const match = value.match(/(\d+)(.+)/);
    if (match === null || match.length !== 3) {
      return false;
    }

    const key = Object.entries(durationMap).find(([_, abbreviations]) => abbreviations?.includes(match[2]))?.[0];
    if (!key) {
      return false;
    }
  }

  return true;
}
```

For config items 'Query timeout' and 'Scrape interval' we never supported multiple durations. For incremental querying, do we support multiple durations?

gtk-grafana replied:
Yeah but just because that function does, if this is the desired functionality then we should be good

bohandley replied:
updated validation for incremental querying! https://github.com/grafana/grafana/pull/66198/commits/b9a0f01e3c9b25a1ac39c33b53a1d5d9f419e350

gtk-grafana commented on public/app/plugins/datasource/prometheus/configuration/Connection.tsx:
```diff
@@ -0,0 +1,104 @@
+      <div className={`${styles.sectionBottomPadding} ${styles.secondaryGrey}`}>
+        For more information on configuring the Grafana Prometheus data source see the{' '}
+        <a
+          style={{ textDecoration: 'underline' }}
```
we've got some inline styles, and some class names, can we make this a bit more consistent?

bohandley replied:
Can do!

bohandley replied:
updated! https://github.com/grafana/grafana/pull/66198/commits/df31e475db0cabd7998f674125a18fc8e51e249e

gtk-grafana commented on public/app/plugins/datasource/prometheus/configuration/Connection.tsx:
```diff
@@ -0,0 +1,104 @@
+      urlTooltip = 'Specify a complete HTTP URL (for example http://your_server:8080)';
+  }
+
+  const validUrlRegex = /^(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?$/;
```
Where did we get this from? Is this currently in use in grafana anywhere? Do we have any additional documentation as to what urls this accepts?

Maybe unit tests for this if it's not already tested

bohandley replied:
This connection component logic was copied from https://github.com/grafana/grafana/blob/39a3d85514553e87bd17e516bcf97373e14ad9f6/packages/grafana-ui/src/components/DataSourceSettings/DataSourceHttpSettings.tsx#L122-L124 which implements this regex... but it does not test it 😆 I will write some unit tests for it.

bohandley replied:
tests added! https://github.com/grafana/grafana/pull/66198/commits/7ef949ccefdbaa1bad51b78a717b009baaba6562

gtk-grafana reviewed (dismissed):
The FE looks good, a few small nits.
Including we've got a bunch of selects that aren't getting the disabled style for provisioned sources, and users can interact and (appear to) change their values.
<img width="722" alt="image" src="https://user-images.githubusercontent.com/109082771/234309381-141e1541-bc63-4967-a599-209111580294.png">

bohandley committed:
Update public/app/plugins/datasource/prometheus/configuration/Connection.tsx

Co-authored-by: Torkel Ödegaard <torkel@grafana.com>

bohandley committed:
fix read only issue for provisioned datasources

bohandley committed:
Merge branch 'bohandley/prometheus-config-overhaul' of https://github.com/grafana/grafana into bohandley/prometheus-config-overhaul

gtk-grafana commented:
Still lots of title case copy, can we switch to sentence casing?
<img width="534" alt="image" src="https://user-images.githubusercontent.com/109082771/234303605-de0901a0-d353-41da-a1c3-34942c41bc90.png">

bohandley commented:
@gtk-grafana @catherineymgui @torkelo 

> Still lots of title case copy, can we switch to sentence casing? <img alt="image" width="534" src="https://user-images.githubusercontent.com/109082771/234303605-de0901a0-d353-41da-a1c3-34942c41bc90.png">

Question for sentence casing in Config headers, the example I was given for headers used title casing and not sentence casing. Which should I use, title or sentence?

<img width="353" alt="Screenshot 2023-04-25 at 11 15 54 AM" src="https://user-images.githubusercontent.com/25674746/234324238-1abd1d31-72a1-4a53-b231-04d7317962ab.png">

catherineymgui commented:
@bohandley 

Sorry, I totally missed this one. Please use sentence casing in all headers and labels as seen in the Figma designs!

bohandley committed:
validate multiple durations for incremental query setting

bohandley commented:
From @torkelo 
> Seems like very very inconsistent amount and form of spacing. Why the huge margins and separator line just for the URL property?

@catherineymgui @josiahg The config overhaul group requested that we have a "bare minimum" section that is separate from all other elements and the requested form of separation was the separator line with padding space of 32px. The bare minimum for Prometheus is a connection URL.  Other datasources require auth as part of the bare minimum and you can see that in the Oracle config. Ultimately I want to use padding instead of margin css in all components but as I am waiting on components being developed by enterprise squad folk I am using margins which allow for the desired space between the `<hr /> tags that are used for line separators (You can't add padding to those).

![Screenshot 2023-04-12 at 15 51 42](https://user-images.githubusercontent.com/25674746/234330779-3ca094ce-cc85-42f2-8645-55d4dd28f504.png)

bohandley committed:
use sentence case for headers

bohandley committed:
use className consistently for styles

torkelo commented:
Name is also required, just looked a bit very strange (especially the line and border under Name). And 32px is more spacing between elements that I have seen we use for anything.

bohandley committed:
add tests for url regex

bohandley committed:
Merge branch 'main' into bohandley/prometheus-config-overhaul

catherineymgui commented:
@torkelo We're only using the 32px spacing between sections (not each field) for enough white space between them. It is a bit tight without the spacing. As Brendan said, there are some components that have to be updated in the 2nd round of redesigns so there's more consistency with the connection and HTTP settings parts.

Fyi - the name field is getting pulled out into an editable header component.

bohandley committed:
remove console logs

bohandley commented:
@catherineymgui 

I had to change the implementation a little and simplify the health check as I found out that that Mimir does not implement the `healthy` api. Considering this and considering 'cloud first', I am moving forward with simplifying the health check and doing what the folks on the Mimir team suggested, simply performing a query and returning the errors.

I suggest any particular information we express by the error messages themselves and we add a generic message. Due to the responses returned and how Prometheus backend returns errors for queries, the way that I could disambiguate the errors and add special messages for each error would be brittle. If there was one perfect route to handle all types and versions for Prometheus, this would work, but sadly that is not the case.

I strongly suggest using this generic error message.
"Your configuration has been saved but there is an error."

This would look like the following:

1. `401 Unauthorized - Your configuration has been saved but there is an error.`
2. `404 Not Found - Your configuration has been saved but there is an error.`
3. `Error querying resource: Get “[http://gobbledygook”:](http://this./-/healthy%E2%80%9D:) dial tcp 143.244.220.150.80: io timeout - Your configuration has been saved but there is an error.`

And for success, this one is perfect 🎉 

`Successfully saved the configuration and queried the Prometheus API.`

bohandley committed:
only use query as healthcheck as the healthy api is not supported by Mimir

bohandley committed:
fix exemplar e2e test

torkelo commented:
@catherineymgui Just the spacing looks so different from any other large form we have, and on a small laptop seems very excessive 
<img width="1264" alt="Screenshot 2023-04-26 at 07 16 34" src="https://user-images.githubusercontent.com/10999/234476705-f7296fa4-afb1-4b23-b34a-f3f602f664e8.png">

Compared to dashboard settings:

<img width="959" alt="Screenshot 2023-04-26 at 07 17 40" src="https://user-images.githubusercontent.com/10999/234476797-99ae9dc6-d433-4829-bda3-2336650d5fd8.png">

Alert contacts 
<img width="780" alt="Screenshot 2023-04-26 at 07 17 05" src="https://user-images.githubusercontent.com/10999/234476968-922691a0-42ee-4b76-ad1c-59b761b05478.png">

bohandley committed:
remove overhaul prop from custom headers setting component

bohandley commented:
@torkelo and @catherineymgui 

Here are three possible ways we can go with the Connection URL styling. Because this is stage 1 work I am inclined to do the first choice because we will return to the styling in stage 2 with better components and it required the least code changes. But, I am open to doing the other two options which make incremental movements towards the UI suggestions based on the research of the config overhaul working group. WDYT?

1. Use the previous DatasourceHttpSettings URL input
  - Pros: Less excessive spacing on small screens
  - Cons: Missing UI parts that add clarity to the Prometheus URL set up (docs link, Prom custom field label), uses legacy form component
<img width="495" alt="Screenshot 2023-04-26 at 10 26 00 AM" src="https://user-images.githubusercontent.com/25674746/234612589-08d04309-88a4-4fe4-b466-f08fef56c3ed.png">

2. Remove all custom space styling (excessive space) around Connection section
  - Pros: could look better on smaller screens, has all Prom clarity details like field label, required star, tooltip with link
  - Cons: the design looks a little off for this style, still looks like we need a little and I would need some UI direction in what space to add if we need it.)
<img width="538" alt="Screenshot 2023-04-26 at 10 38 15 AM" src="https://user-images.githubusercontent.com/25674746/234614969-da213ad0-f892-463f-82d0-35d69d372207.png">

3. Pass the Connection label input and link to docs to set on the tooltip into the DatasourceHttpSettings
  - Pros: adds clarity to setting up datasource, custom prom label and link to docs in the tooltip (interactive)
  - Cons: Can't show the required star on the label because DatasourceHttpSettings is using the legacy form field label + input.
<img width="506" alt="Screenshot 2023-04-26 at 11 45 04 AM" src="https://user-images.githubusercontent.com/25674746/234630737-6e25e15f-525a-40d9-9e3b-19c720c0ebe4.png">

bohandley committed:
remove connection section and use DatasourceHttpSettings connection with custom label and interactive tooltip

bohandley commented:
After speaking with @catherineymgui, we compromised for the Connection styling @torkelo pointed out with this choice, choice three from the above comment. Please let me know if there are concerns.

<img width="528" alt="Screenshot 2023-04-26 at 3 28 41 PM" src="https://user-images.githubusercontent.com/25674746/234682998-7f360da7-a31e-4057-8c56-daf0baa8b79e.png">

torkelo commented:
@bohandley look ok 👍

bohandley committed:
add spaces back

bohandley committed:
spaces

kylebrandt commented on pkg/tsdb/prometheus/healthcheck.go:
```diff
@@ -0,0 +1,93 @@
+	})
+
+	if err != nil {
+		return getHealthCheckMessage(logger, "Your configuration has been saved but there is an error.", err)
```
Since this part doesn't do any saving, and just health check, I think it should just be an error message regarding the health check.

gtk-grafana reviewed (approved):
Looks good, thanks for all your hard work on this Brendan!

bohandley commented on pkg/tsdb/prometheus/healthcheck.go:
```diff
@@ -0,0 +1,93 @@
+)
+
+const (
+	refID = "__healthcheck__"
```
Adding this for work here https://github.com/grafana/observability-metrics-squad/issues/116. Thank you Kyle!

torkelo reviewed (commented):
Dam this was merged before I review & approve :(

will open a Fix PR

bohandley commented:
My apologies @torkelo, I thought the last reviews were the UI ones. I was incorrect and merged this too quickly. I am sorry.

Thank you for opening the fix for the DatasourceHttpSettings, I am reviewing that now.

bohandley commented on public/app/plugins/datasource/prometheus/datasource.tsx:
```diff
@@ -1051,44 +1051,6 @@ export class PrometheusDatasource
     );
   }
 
-  async testDatasource() {
```
@sasklacz We removed it in this PR

torkelo commented:
@bohandley No worries, should not have to feel you need to wait for approval from me. Was it my mistake on being late with a review (or I should have marked the first review as request changes)

bohandley commented on pkg/tsdb/prometheus/querydata/request.go:
```diff
@@ -179,6 +179,13 @@ func (s *QueryData) instantQuery(ctx context.Context, c *client.Client, q *model
 		}
 	}
 
+	// This is only for health check fall back scenario
```
@AndrejKiri here PR and the spot where we return errors for health check
